File: 프로그래머스/3/42898. 등굣길/등굣길.cpp
#include <string>
#include <vector>
#include <queue>
#include <tuple>
#include <string.h>

using namespace std;

int dy[] = {0, 1};
int dx[] = {1, 0};

int visited[104][104] = {0,};
int map[104][104] = {0,};
long long dp[104][104] = {0,};

int ny, nx, y, x, mod = 1000000007;

int solution(int m, int n, vector<vector<int>> p) {
    queue<pair<int, int>> q;

    for (int i = 0; i < p.size(); i++) {
        map[p[i][0] - 1][p[i][1] - 1] = -1;
    }

    q.push({0, 0});
    visited[0][0] = 1;
    dp[0][0] = 1;

    while (!q.empty()) {
        tie(y, x) = q.front(), q.pop();

        if (y == m - 1 && x == n - 1) break;

        for (int i = 0; i < 2; i++) {
            int ny = y + dy[i];
            int nx = x + dx[i];
            if (ny < 0 || ny >= m || nx < 0 || nx >= n || map[ny][nx] == -1) continue;
            dp[ny][nx] += dp[y][x] % mod;
            if (visited[ny][nx]) continue;
            q.push({ny, nx});
            visited[ny][nx] = 1;
        }
    }

    return dp[m - 1][n - 1] % mod;
}

File: 프로그래머스/3/42898. 등굣길/등굣길.java
class Solution {
    
    static int[] dx = {0, 1};
    static int[] dy = {1, 0};
    
    public int solution(int m, int n, int[][] puddles) {
        int[][] map = new int[n][m];
        for(var i : puddles) {
            map[i[1] - 1][i[0] - 1] = -1;
        }
        int[][] dp = new int[n][m];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                dp[i][j] = -1;
            }
        }
        
        return this.dfs(m, n, 0, 0, map, dp);
    }
    
    private int dfs(int m, int n, int i, int j, int[][] map, int[][] dp) {
        if (i == n - 1 && j == m - 1) {
            return 1;
        }
        
        if (dp[i][j] != -1) {
            return dp[i][j];
        }

        dp[i][j] = 0;
        
        for (int d = 0; d < 2; d++) {
            int nx = i + dx[d];
            int ny = j + dy[d];
            if (0 <= nx && nx < map.length && 0 <= ny && ny < map[0].length && map[nx][ny] != -1) {
                dp[i][j] = (dp[i][j] + dfs(m, n, nx, ny, map, dp)) % 1_000_000_007;
            }
        }
        
        return dp[i][j];
    }
}